File: WeLegends/andiag-mvp/src/main/java/es/coru/andiag/andiag_mvp/BaseFragment.java
package es.coru.andiag.andiag_mvp;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;

import es.coru.andiag.andiag_mvp.base.BaseFragmentPresenter;
import es.coru.andiag.andiag_mvp.interfaces.BaseFragmentView;


/**
 * Created by Canalejas on 11/12/2016.
 */

public abstract class BaseFragment<P extends BaseFragmentPresenter> extends Fragment implements BaseFragmentView {
    private final static String TAG = BaseFragment.class.getSimpleName();

    protected Context mParentContext;
    private P mPresenter;

    protected void setPresenter(P presenter) {
        mPresenter = presenter;
    }

    protected P getPresenter() {
        return mPresenter;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.mParentContext = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (mPresenter == null) Log.e(TAG, "Null Presenter: Initialize it on setPresenter method");
        this.mPresenter.attach(this, this.mParentContext);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.mPresenter.detach();
        this.mPresenter = null;
        this.mParentContext = null;
    }

}

File: WeLegends/app/src/main/java/es/coru/andiag/welegends/common/BaseFragment.kt
package es.coru.andiag.welegends.common

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import butterknife.Unbinder

/**
 * Created by andyq on 09/12/2016.
 */
abstract class BaseFragment<P : BasePresenter<*> > () : Fragment() {

    lateinit var activity: Activity

    lateinit internal var unbinder: Unbinder
    var presenter: P? = null
        protected set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.presenter = this.addPresenter()
        this.onPresenterCreated(this.presenter!!)
    }

    override fun onAttach(activity: Context?) {
        super.onAttach(activity)
        this.activity = activity as Activity
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(fragmentLayout, container, false)
        unbinder = ButterKnife.bind(this, view)
        return view
    }

    protected abstract val fragmentLayout: Int

    abstract fun setupViews()

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupViews()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        unbinder.unbind()
        if (this.presenter != null) {
            (this.presenter as BasePresenter<*>).detach()
        }
    }

    abstract fun addPresenter(): P

    abstract fun onPresenterCreated(var1: P)
}
